core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala

```diff
@@ -41,7 +40,41 @@ class TransactionMarkerChannel(interBrokerListenerName: ListenerName,
   // we need the metadataPartition so we can clean up when Transaction Log partitions emigrate
   case class PendingTxnKey(metadataPartition: Int, producerId: Long)
 
-  private val brokerStateMap: concurrent.Map[Int, DestinationBrokerAndQueuedMarkers] = concurrent.TrieMap.empty[Int, DestinationBrokerAndQueuedMarkers]
+  class BrokerRequestQueue(private var destination: Node) {
+
+    // keep track of the requests per metadata partition so we can easily clear the queue
+    // during partition emigration
+    private val requestsPerMetadataPartition: concurrent.Map[Int, BlockingQueue[TxnMarkerEntry]]
+      = concurrent.TrieMap.empty[Int, BlockingQueue[TxnMarkerEntry]]
+
+    def removeRequestsForPartition(partition: Int): Unit = {
+      requestsPerMetadataPartition.remove(partition)
+    }
+
+    def maybeUpdateNode(node: Node): Unit = {
+      if (!destination.equals(node))
+        destination = node
+    }
+
+    def addRequests(metadataPartition: Int, txnMarkerEntry: TxnMarkerEntry): Unit ={
+      val queue = requestsPerMetadataPartition.getOrElseUpdate(metadataPartition, new LinkedBlockingQueue[TxnMarkerEntry]())
+      queue.add(txnMarkerEntry)
+    }
+
+    def eachMetadataPartition[B](f:(Int, BlockingQueue[TxnMarkerEntry]) => B): mutable.Iterable[B] =
+      requestsPerMetadataPartition.filter{case(_, queue:BlockingQueue[TxnMarkerEntry]) => !queue.isEmpty}
+        .map{case(partition:Int, queue:BlockingQueue[TxnMarkerEntry]) => f(partition, queue)}
+
+
+    def node: Node = destination
+
+    def totalQueuedRequests(): Int =
+      requestsPerMetadataPartition.map {case(_, queue:BlockingQueue[TxnMarkerEntry]) => queue.size()}
+        .sum
+
+  }
+
+  private val brokerStateMap: concurrent.Map[Int, BrokerRequestQueue] = concurrent.TrieMap.empty[Int, BrokerRequestQueue]
   private val pendingTxnMap: concurrent.Map[PendingTxnKey, TransactionMetadata] = concurrent.TrieMap.empty[PendingTxnKey, TransactionMetadata]
```

Comment:
Yeah I'm aware of that, and hence was thinking whether we can change `TransactionMarkerRequestCompletionHandler`'s constructor parameters to augment the `CoordinatorEpochAndMarkers` to encapsulate the `transactionalId` information as well. But thinking it one more time it may not be worthwhile to remove this additional map. So nvm.